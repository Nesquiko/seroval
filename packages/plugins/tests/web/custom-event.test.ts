import {
  crossSerialize,
  crossSerializeAsync,
  crossSerializeStream,
  deserialize,
  fromCrossJSON,
  fromJSON,
  serialize,
  serializeAsync,
  toCrossJSON,
  toCrossJSONAsync,
  toCrossJSONStream,
  toJSON,
  toJSONAsync,
} from 'seroval';
import { describe, expect, it } from 'vitest';
import CustomEventPlugin from '../../web/custom-event';

const EXAMPLE_EVENT_TYPE = 'example';
const EXAMPLE_DETAIL: Record<string, unknown> = {};
EXAMPLE_DETAIL.self = EXAMPLE_DETAIL;
const EXAMPLE_OPTIONS: CustomEventInit = { detail: EXAMPLE_DETAIL };
const EXAMPLE = new CustomEvent(EXAMPLE_EVENT_TYPE, EXAMPLE_OPTIONS);

describe('CustomEvent', () => {
  describe('serialize', () => {
    it('supports CustomEvent', () => {
      const result = serialize(EXAMPLE, {
        plugins: [CustomEventPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = deserialize<typeof EXAMPLE>(result);
      expect(back).toBeInstanceOf(CustomEvent);
    });
  });
  describe('serializeAsync', () => {
    it('supports CustomEvent', async () => {
      const result = await serializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [CustomEventPlugin],
      });
      expect(result).toMatchSnapshot();
      const back = await deserialize<Promise<typeof EXAMPLE>>(result);
      expect(back).toBeInstanceOf(CustomEvent);
    });
  });
  describe('toJSON', () => {
    it('supports CustomEvent', () => {
      const result = toJSON(EXAMPLE, {
        plugins: [CustomEventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromJSON<typeof EXAMPLE>(result, {
        plugins: [CustomEventPlugin],
      });
      expect(back).toBeInstanceOf(CustomEvent);
    });
  });
  describe('toJSONAsync', () => {
    it('supports CustomEvent', async () => {
      const example = Promise.resolve(EXAMPLE);
      const result = await toJSONAsync(example, {
        plugins: [CustomEventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromJSON<typeof example>(result, {
        plugins: [CustomEventPlugin],
      });
      expect(back).toBeInstanceOf(CustomEvent);
    });
  });
  describe('crossSerialize', () => {
    it('supports CustomEvent', () => {
      const result = crossSerialize(EXAMPLE, {
        plugins: [CustomEventPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports CustomEvent', () => {
        const result = crossSerialize(EXAMPLE, {
          plugins: [CustomEventPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeAsync', () => {
    it('supports CustomEvent', async () => {
      const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
        plugins: [CustomEventPlugin],
      });
      expect(result).toMatchSnapshot();
    });
    describe('scoped', () => {
      it('supports CustomEvent', async () => {
        const result = await crossSerializeAsync(Promise.resolve(EXAMPLE), {
          plugins: [CustomEventPlugin],
          scopeId: 'example',
        });
        expect(result).toMatchSnapshot();
      });
    });
  });
  describe('crossSerializeStream', () => {
    it('supports CustomEvent', async () =>
      new Promise<void>((resolve, reject) => {
        crossSerializeStream(Promise.resolve(EXAMPLE), {
          plugins: [CustomEventPlugin],
          onSerialize(data) {
            expect(data).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
    describe('scoped', () => {
      it('supports CustomEvent', async () =>
        new Promise<void>((resolve, reject) => {
          crossSerializeStream(Promise.resolve(EXAMPLE), {
            plugins: [CustomEventPlugin],
            scopeId: 'example',
            onSerialize(data) {
              expect(data).toMatchSnapshot();
            },
            onDone() {
              resolve();
            },
            onError(error) {
              reject(error);
            },
          });
        }));
    });
  });
  describe('toCrossJSON', () => {
    it('supports CustomEvent', () => {
      const result = toCrossJSON(EXAMPLE, {
        plugins: [CustomEventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = fromCrossJSON<typeof EXAMPLE>(result, {
        plugins: [CustomEventPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(CustomEvent);
    });
  });
  describe('toCrossJSONAsync', () => {
    it('supports CustomEvent', async () => {
      const example = Promise.resolve(EXAMPLE);
      const result = await toCrossJSONAsync(example, {
        plugins: [CustomEventPlugin],
      });
      expect(JSON.stringify(result)).toMatchSnapshot();
      const back = await fromCrossJSON<typeof example>(result, {
        plugins: [CustomEventPlugin],
        refs: new Map(),
      });
      expect(back).toBeInstanceOf(CustomEvent);
    });
  });
  describe('toCrossJSONStream', () => {
    it('supports CustomEvent', async () =>
      new Promise<void>((resolve, reject) => {
        toCrossJSONStream(Promise.resolve(EXAMPLE), {
          plugins: [CustomEventPlugin],
          onParse(data) {
            expect(JSON.stringify(data)).toMatchSnapshot();
          },
          onDone() {
            resolve();
          },
          onError(error) {
            reject(error);
          },
        });
      }));
  });
});
