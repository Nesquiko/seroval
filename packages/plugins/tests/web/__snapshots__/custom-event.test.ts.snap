// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomEvent > crossSerialize > scoped > supports CustomEvent 1`] = `"($R=>($R[0]=new CustomEvent("example",$R[1]={detail:$R[2]={},bubbles:!1,cancelable:!1,composed:!1}),$R[2].self=$R[2],$R[0]))($R["example"])"`;

exports[`CustomEvent > crossSerialize > supports CustomEvent 1`] = `"($R[0]=new CustomEvent("example",$R[1]={detail:$R[2]={},bubbles:!1,cancelable:!1,composed:!1}),$R[2].self=$R[2],$R[0])"`;

exports[`CustomEvent > crossSerializeAsync > scoped > supports CustomEvent 1`] = `"($R=>($R[0]=Promise.resolve($R[1]=new CustomEvent("example",$R[2]={detail:$R[3]={},bubbles:!1,cancelable:!1,composed:!1})),$R[3].self=$R[3],$R[0]))($R["example"])"`;

exports[`CustomEvent > crossSerializeAsync > supports CustomEvent 1`] = `"($R[0]=Promise.resolve($R[1]=new CustomEvent("example",$R[2]={detail:$R[3]={},bubbles:!1,cancelable:!1,composed:!1})),$R[3].self=$R[3],$R[0])"`;

exports[`CustomEvent > crossSerializeStream > scoped > supports CustomEvent 1`] = `"($R=>$R[0]=($R[2]=r=>(r.p=new Promise((s,f)=>{r.s=s,r.f=f})))($R[1]={p:0,s:0,f:0}))($R["example"])"`;

exports[`CustomEvent > crossSerializeStream > scoped > supports CustomEvent 2`] = `"($R=>(($R[6]=(r,d)=>{r.s(d),r.p.s=1,r.p.v=d})($R[1],$R[3]=new CustomEvent("example",$R[4]={detail:$R[5]={},bubbles:!1,cancelable:!1,composed:!1})),$R[5].self=$R[5],$R[1]))($R["example"])"`;

exports[`CustomEvent > crossSerializeStream > supports CustomEvent 1`] = `"$R[0]=($R[2]=r=>(r.p=new Promise((s,f)=>{r.s=s,r.f=f})))($R[1]={p:0,s:0,f:0})"`;

exports[`CustomEvent > crossSerializeStream > supports CustomEvent 2`] = `"(($R[6]=(r,d)=>{r.s(d),r.p.s=1,r.p.v=d})($R[1],$R[3]=new CustomEvent("example",$R[4]={detail:$R[5]={},bubbles:!1,cancelable:!1,composed:!1})),$R[5].self=$R[5],$R[1])"`;

exports[`CustomEvent > serialize > supports CustomEvent 1`] = `"((h,j)=>(j=new CustomEvent("example",{detail:h={},bubbles:!1,cancelable:!1,composed:!1}),h.self=h,j))()"`;

exports[`CustomEvent > serializeAsync > supports CustomEvent 1`] = `"((h,j)=>(j=Promise.resolve(new CustomEvent("example",{detail:h={},bubbles:!1,cancelable:!1,composed:!1})),h.self=h,j))()"`;

exports[`CustomEvent > toCrossJSON > supports CustomEvent 1`] = `"{"t":25,"i":0,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":1,"p":{"k":["detail","bubbles","cancelable","composed"],"v":[{"t":10,"i":2,"p":{"k":["self"],"v":[{"t":4,"i":2}],"s":1},"o":0},{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":4},"o":0}},"c":"seroval-plugins/web/CustomEvent"}"`;

exports[`CustomEvent > toCrossJSONAsync > supports CustomEvent 1`] = `"{"t":12,"i":0,"s":1,"f":{"t":25,"i":1,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":2,"p":{"k":["detail","bubbles","cancelable","composed"],"v":[{"t":10,"i":3,"p":{"k":["self"],"v":[{"t":4,"i":3}],"s":1},"o":0},{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":4},"o":0}},"c":"seroval-plugins/web/CustomEvent"}}"`;

exports[`CustomEvent > toCrossJSONStream > supports CustomEvent 1`] = `"{"t":22,"i":0,"s":1,"f":{"t":26,"i":2,"s":1}}"`;

exports[`CustomEvent > toCrossJSONStream > supports CustomEvent 2`] = `"{"t":23,"i":1,"a":[{"t":26,"i":6,"s":2},{"t":25,"i":3,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":4,"p":{"k":["detail","bubbles","cancelable","composed"],"v":[{"t":10,"i":5,"p":{"k":["self"],"v":[{"t":4,"i":5}],"s":1},"o":0},{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":4},"o":0}},"c":"seroval-plugins/web/CustomEvent"}]}"`;

exports[`CustomEvent > toJSON > supports CustomEvent 1`] = `"{"t":{"t":25,"i":0,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":1,"p":{"k":["detail","bubbles","cancelable","composed"],"v":[{"t":10,"i":2,"p":{"k":["self"],"v":[{"t":4,"i":2}],"s":1},"o":0},{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":4},"o":0}},"c":"seroval-plugins/web/CustomEvent"},"f":31,"m":[2]}"`;

exports[`CustomEvent > toJSONAsync > supports CustomEvent 1`] = `"{"t":{"t":12,"i":0,"s":1,"f":{"t":25,"i":1,"s":{"type":{"t":1,"s":"example"},"options":{"t":10,"i":2,"p":{"k":["detail","bubbles","cancelable","composed"],"v":[{"t":10,"i":3,"p":{"k":["self"],"v":[{"t":4,"i":3}],"s":1},"o":0},{"t":2,"s":3},{"t":2,"s":3},{"t":2,"s":3}],"s":4},"o":0}},"c":"seroval-plugins/web/CustomEvent"}},"f":31,"m":[3]}"`;
